### PR TITLE
Ammo Variety Mod

### PR DESCRIPTION
**Pull request by DangerNoodle**
This moves the content of two pull requests to a mod, due to general consensus against further making ammo type overlap more complex in mainline.

Content from https://github.com/CleverRaven/Cataclysm-DDA/pull/18972, by @BorkBorkGoesTheCode:
- Addition of 10mm Auto as a separate ammunition type.
- Allowing the S&W 610 to chamber both 10mm Auto and .40 S&W.
- Editing ammunition conversion kits to be used on 10mm firearms.

Content from https://github.com/CleverRaven/Cataclysm-DDA/pull/18986:
- Implementing .45 Long Colt and .410 bore ammunition.
- Implementing the Colt Single Action Army, Colt Lightning, Bond Arms Derringer, .410 youth shotgun, .410 pipe shotgun, and M6 Aircrew Survival Weapon.

New changes from prior pull requests:
- Streamlined the editing of 10mm ammo, .40 ammo, and conversion kits to not require full overrides.
- Altered the ammo group placement of .45 LC guns and ammo.
- Allowing the altered conversion kits to additionally be used on .45 LC ammunition.
- Implementing a .410 equivalent to scrap shot, and .410 conversion kit.

Future plans include separation of .38 Special from .38 Super, separation of .45 ACP from .45 Super, separation of .270 Winchester from .30-06, and implementation or .357 Magnum.

One complication is presently apparent. Due to ammo type overlap, 10mm Auto firearms can automatically also chambered in .40 S&W. While this is acceptable for the revolver presented as an example, they would not be interchangeable in a semi-automatic firearm. This can be resolved by the proposed "overlap at gun level" addition suggested by @Mugling.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Could this be included in one of the "more guns" mods?
We don't really need 3+ of "more gun related things" mods as people who want extra gun variety will pick one of the mods anyway.

While (at least) one of the mods is named after the author, I don't think the author is still around so there is no need to keep that name.

**Comment by BorkBorkGoesTheCode** [comment]
Are you talking about the idea of uniting Icecoon's mod and DeadLeaves?

**Comment by DangerNoodle** [comment]
That might be an option. Alternatively, a partial streamlining can occur via combining Icecoon's Weapons Pack and Fictional Guns. Both of these are focused on (as the latter's name implies) fictional weaponry, whereas there currently are not any mods (barring Artyom's unofficial mod and that one nazi mod) focused on real-world firearms.

More firearms in general might be an acceptable thing to combine in one mod, it just depends on what the consensus is regarding real and fictional weapons.

**Comment by BorkBorkGoesTheCode** [comment]
Since this mod is more realism-based, I favor having this mod independent from the others, but I think combining Icecoon's and DeadLeaves mods is a good idea. What do you think @DeadLeaves?

**Comment by mugling** [comment]
Mods (particularly those that add complex items) impose a significant cost on the developers. Mods that are included in core should have:

* reasonably broad appeal
* an active maintainer
* acceptable code standards

The latter point eases maintenance burden and whilst it can be relaxed for existing mods we should try and void JSON-soup in new mods.

This PR fulfills all those criteria. In particular I can see a *'realistic guns'* mod having sufficient appeal to be worthy of the maintenance burden and the JSON is well structured and formatted which should help maintenance (you should consider adding this to `json_whitelist` as I suspect it passes).

That said we don't want a proliferation of gun mods in `core` as this dilutes down both broad appeal and leaves us lacking sufficient maintainers. We need a *'generic guns'* mod and probably one other *'fictional guns'* mod. Combining the existing fictional mods is therefore a good idea.

File: data/mods/Ammo_Variety/modinfo.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "type": "MOD_INFO",
+    "mod-type": "SUPPLEMENTAL",
+    "ident": "Ammo_Variety",
+    "name": "Expanded Ammunition Variety",
```
Comment (mugling) [inline data/mods/Ammo_Variety/modinfo.json]: Needs better name, for example `expanded realistic guns`

**Comment by DangerNoodle** [comment]
Regarding the mod name, I was unsure what to use for that. Expanded Realistic Guns, More Guns, etc. Expanded Realistic Guns might work, unless @BorkBorkGoesTheCode has any ideas?

**Comment by mugling** [comment]
The obvious split is 'Extended Realistic Guns' and 'Extended Fictional Guns' or possibly just 'Fictional Guns'

**Comment by BorkBorkGoesTheCode** [comment]
From the folder name of Icecoons mod, it looks like it was Extended Weapons mod

**Comment by DangerNoodle** [comment]
Extended Realistic Guns might work. Or addressing my earlier concern, Extended Real-world Guns might carry different implications.

**Comment by mugling** [comment]
I'd go with the more concise *Extended Realistic Guns*. I'm working on the mod-loading code at present so it would be good to get this (and the DeadLeaves PR) merged reasonably soon.

**Comment by DangerNoodle** [comment]
I see. Very well then. As stated, I am still concerned about implications and expectations of labeling a mod "realistic" but it is not essential if no better name can be devised.

**Comment by mugling** [comment]
It's trivial to change the mod name later.

You should use `realguns` as the identifier though as that works for either

**Comment by DangerNoodle** [comment]
Very well then. Should I change the folder name as well, or just the ID and name in modinfo? The latter will be faster due to needing to sync client on my desktop.

**Comment by mugling** [comment]
The folder name should match the ident

**Comment by DangerNoodle** [comment]
As I suspected, hence went with changing via client.

File: data/mods/realguns/ammo.json [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+      "dispersion": 1.1
+    },
+    "extend": { "effects": [ "RECYCLED" ] },
+    "delete": {
+      "ammo_type": [ "40" ],
+      "effects": [ "NEVER_MISFIRES" ]
+    }
+  },
+  {
+    "id": "45lc_casing",
```
Comment (mugling) [inline data/mods/realguns/ammo.json]: See `data/json/items/generic/casing.json` for a better definition of cartridge casings

**Comment by mugling** [comment]
Otherwise looks good to go

File: data/mods/realguns/ammo.json (hunk 2 above)
Comment (DangerNoodle) [inline data/mods/realguns/ammo.json]: When was this added? I am assuming recently, as it was not in this branch,

**Comment by Rivet-the-Zombie** [comment]
Apparently I'm in the minority here, but I'd much rather see this stuff in the mainline.

We don't use the 'it would require some people learning a new piece of information' excuse to dumb down vehicles in this game, so I see no reason why we should we apply that logic to firearms.

**Comment by Rivet-the-Zombie** [comment]
Apparently I'm in the minority here, but I'd much rather see this stuff in the mainline.

We've not used the 'it would require some people learning a new piece of information' excuse to dumb down vehicles in this game, so I see no reason why we should we apply that logic to firearms.

**Comment by Rivet-the-Zombie** [comment]
Apparently I'm in the minority here, but I'd much rather see this stuff in the mainline.

We've not used the 'it would require some people learning a new piece of information' excuse to dumb down vehicles in this game, so I see no reason why we should start applying that logic to firearms.

**Comment by Rivet-the-Zombie** [comment]
Apparently I'm in the minority here, but I'd much rather see this stuff in the mainline.

We've not used the 'it would require some people to learn a new piece of information' excuse to dumb down vehicles in this game, so I see no reason why we should start applying that logic to firearms.

**Comment by mugling** [comment]
It doesn't add anything in terms of gameplay as a payoff for the extra complexity. The requirement for mainline being broad majority and the requirement for a mod being significant minority.

**Comment by DangerNoodle** [comment]
I am fine with it being implemented via either method.

**Comment by mugling** [comment]
```
$ ./cataclysm-tiles --check-mods realguns                                                                                                        
Checking mod Extended Realistic Guns [realguns]
src/requirements.cpp:306 [void component::check_consistency(const std::string &) const] 45lc_casing in ammo_45lc is not a valid item template
src/item_factory.cpp:831 [void Item_factory::check_definitions() const] warnings for type 45_long_colt:
invalid casing property 45lc_casing

src/item_factory.cpp:831 [void Item_factory::check_definitions() const] warnings for type reloaded_45_long_colt:
invalid casing property 45lc_casing

src/item_group.cpp:116 [virtual void Single_item_creator::check_consistency() const] item id 45lc_casing is unknown
```

**Comment by DangerNoodle** [comment]
Ah, apologies for that, typo introduced when editing the new style of casing.

**Comment by DangerNoodle** [comment]
Thank you for the merge. Now if I recall, @BorkBorkGoesTheCode mentioned they had .357 Magnum content they could add?

**Comment by BorkBorkGoesTheCode** [comment]
Hi, @DangerNoodle . I have nearly finished uploading .357 content to GitHub, and I will PR within 2 days if possible.

**Comment by DangerNoodle** [comment]
Ah, will be nice to see.

**Pull request merged**